README.md
Private Prediction
========

This repository contains code that can be used to reproduce the experimental
results presented in the paper:

- L.J.P. van der Maaten* and A.Y. Hannun*. **[The Trade-Offs of Private Prediction](https://arxiv.org/abs/2007.05089)**. arXiv 2007.05089, 2020.

# Installation

The code requires Python 3.5+, [PyTorch 1.5.0+](https://github.com/pytorch/pytorch),
[torchvision 0.6.0+](https://github.com/pytorch/vision), and
[visdom](https://github.com/facebookresearch/visdom) (optional).
It also uses parts of [TensorFlow Privacy](https://github.com/tensorflow/privacy)
and [pytorch_resnet_cifar10](https://github.com/akamaster/pytorch_resnet_cifar10).

Presuming you have installed [Anaconda](https://docs.anaconda.com/anaconda/install/),
you can install all the dependencies via:
```
conda install -c pytorch pytorch torchvision
pip install visdom
python install.py
```

# Usage

All experiments can be run via the `private_prediction_experiment.py` script.
For example, to train and test a linear model on the MNIST dataset using loss
perturbation with privacy loss 1.0, you can use the following command:

```
python private_prediction_experiment.py \
    --dataset mnist \
    --method loss_perturbation \
    --epsilon 1.0
```

The following input arguments can be used to change the model, private prediction
method, and privacy loss:

- `--model`: the model used can be `linear` (default) or `resnet{20,32,44,56,110,1202}`
- `--method`: private prediction method can be `subsagg` (default), `loss_perturbation`, `{model,logit}_sensitivity`, or `dpsgd`
- `--epsilon`: privacy loss value for predictions (default = infinity)
- `--delta`: privacy failure probability for predictions (default = 0.0)
- `--inference_budget`: number of inferences to support (default = -1 to try many values)
- `--weight_decay`: L2-regularization parameter (default = 0.0; set to -1 to cross-validate)

The following input arguments can be used to change details of the optimizer:

- `--optimizer`: optimizer used can be `lbfgs` (default) or `sgd`
- `--num_epochs`: number of training epochs (default = 100)
- `--batch_size`: batch size for SGD optimization (default = 32)
- `--learning_rate`: initial learning rate for SGD optimization (default = 1.0)

The following input arguments alter hyperparameters of specific private prediction methods:

- `--num_models`: number of models in subsample-and-aggregate method (default = 32)
- `--noise_dist`: noise distribution used in sensitivity methods can be `sqrt_gaussian` (default), `laplacian`, `gaussian`, `advanced_gaussian`
- `--clip`: gradient clipping value for DP-SGD (default = 1e-1; set to -1 to cross-validate)
- `--use_lr_scheduler`: use learning rate reduction (for DP-SGD)

The following input arguments alter the dataset used for experimentation:

- `--dataset`: the dataset used can be `mnist` (default), `mnist1m`, `cifar10`, or `cifar100`
- `--num_samples`: number of training samples to use (default: all)
- `--num_classes`: number of classes to use (default: all)
- `--pca_dims`: number of PCA dimensions for data (default: PCA not used)

The following input arguments alter other system properties:

- `--device`: compute device can be `cpu` (default) or `gpu`
- `--visdom`: visdom server for learning curves (default = localhost)
- `--num_repetitions`: number of times to repeat experiment (default = 10)
- `--data_folder`: folder in which to store dataset for re-use
- `--result_file`: file in which to write experimental results (default: unused)

# Using the MNIST-1M Dataset

The MNIST-1M dataset used in the paper is not directly available for download,
but can be generated using [InfiniMNIST](https://leon.bottou.org/projects/infimnist).

Download [InfiniMNIST](https://leon.bottou.org/projects/infimnist) and run:
```
make
mkdir /tmp/infinimnist
infimnist patterns 70000 1069999 > /tmp/infinimnist/mnist1m-images-idx3-ubyte
infimnist labels 70000 1069999 > /tmp/infinimnist/mnist1m-labels-idx1-ubyte
infimnist patterns 0 9999 > t10k-images-idx3-ubyte
infimnist labels 0 9999 > t10k-labels-idx1-ubyte
```

Now, you should be able to run experiments on the MNIST-1M dataset, for example:
```
python private_prediction_experiment.py \
    --dataset mnist1m \
    --num_samples 100000 \
    --method loss_perturbation \
    --epsilon 1.0 \
    --data_folder /tmp/infinimnist
```

# Citing this Repository

If you use the code in this repository, please cite the corresponding paper:

- L.J.P. van der Maaten* and A.Y. Hannun*. **[The Trade-Offs of Private Prediction](https://arxiv.org/abs/2007.05089)**. arXiv 2007.05089, 2020.

# License

This code is released under a CC-BY-NC 4.0 license. Please see the [LICENSE](LICENSE) file for more information.

Please review Facebook Open Source [Terms of Use](https://opensource.facebook.com/legal/terms) and [Privacy Policy](https://opensource.facebook.com/legal/privacy).


CHANGELOG.md
June 28, 2020

### Initial release

* Implements code reproducing all experiments described in the paper.


util.py
#!/usr/bin/env python3

# Copyright (c) Facebook, Inc. and its affiliates.
# All rights reserved.

# This source code is licensed under the license found in the
# LICENSE file in the root directory of this source tree.

import os
import torch


def learning_curve(visualizer, idx, value, window=None, dtype=torch.double, title=""):
    """
    Appends new value to learning curve, creating new curve if none exists.
    """
    opts = {
        "title": title,
        "xlabel": "Epoch",
        "ylabel": "Loss value",
    }
    window = visualizer.line(value.view(value.nelement(), 1), idx,
                             update=None if window is None else "append",
                             opts=opts,
                             win=window,
                             env="shared-model")
    return window


def binary_accuracy(predictions, targets):
    """
    Measures accuracy of `predictions` for binary classification by comparing
    them with given `targets`.
    """
    accuracy = ((predictions > 0.) == (targets == 1)).sum() / float(targets.nelement())
    return accuracy


def accuracy(predictions, targets):
    """
    Measures accuracy of predictions for multi-way classification. Inputs are
    expected to be two matrices with rows containing one-hot or probability
    vectors of length K, `vector1` and `vector2`.
    """
    _, predictions = predictions.topk(1, dim=1, largest=True, sorted=True)
    _, targets = targets.topk(1, dim=1, largest=True, sorted=True)
    accuracy = predictions.eq(targets).sum() / float(targets.size(0))
    return accuracy


def binary_search(func, constraint, minimum, maximum, tol=1e-5):
    """
    Performs binary search on monotonically increasing function `func` between
    `minimum` and `maximum` to find the maximum value for which the function's
    output satisfies the specified `constraint` (which is a binary function).
    Returns maximum value `x` at which `constraint(func(x))` is `True`.

    The function takes an optional parameter specifying the tolerance `tol`.
    """
    assert constraint(func(minimum)), "constraint on function must hold at minimum"

    # evaluate function at maximum:
    if constraint(func(maximum)):
        return maximum

    # perform the binary search:
    while maximum - minimum > tol:
        midpoint = (minimum + maximum) / 2.
        if constraint(func(midpoint)):
            minimum = midpoint
        else:
            maximum = midpoint

    # return value:
    return minimum


noise.py
#!/usr/bin/env python3

# Copyright (c) Facebook, Inc. and its affiliates.
# All rights reserved.

# This source code is licensed under the license found in the
# LICENSE file in the root directory of this source tree.

import torch


def categorical(probabilities, device=None):
    """
    Sample from a categorical distribution with specified `probabilities`.
    """
    sample = torch.distributions.categorical.Categorical(probs=probabilities).sample()
    if device is not None:
        sample = sample.to(device=device)
    return sample


def gaussian(mean, precision, device=None):
    """
    Samples from a Gaussian distribution with the specified `mean` and `precision`.
    """
    sample = torch.distributions.normal.Normal(mean, 1. / precision).sample()
    if device is not None:
        sample = sample.to(device=device)
    return sample


def laplacian(mean, precision, device=None):
    """
    Samples from a Laplace distribution with the specified `mean` and `precision`.
    """
    sample = torch.distributions.laplace.Laplace(mean, 1. / precision).sample()
    if device is not None:
        sample = sample.to(device=device)
    return sample


def sqrt_gaussian(mean, precision, device=None):
    """
    Samples from a square-root Gaussian distribution with the specified
    DxN_1xN_2x...xN_n  tensor `mean` and tensor `precision` which should be a
    scalar or have shape N_1xN_2x...xN_n. The zero-mean square-root Gaussian
    distribution is a multi-variate distribution probability density function:

    p(x) \propto exp(-precision * ||x||_2) with x \in \mathbb{R}^D

    Each column in the output corresponds to a single D-dimensional sample.
    """  # noqa: W905
    assert isinstance(precision, float) or precision.nelement() == 1 \
        or precision.shape == mean.shape[1:], "Invalid shape for precision."

    # sample directions as unit-norm Gaussian vectors:
    direction = torch.randn(mean.size(), device=mean.device)  # samples from Gaussian
    norm = torch.norm(direction, dim=0, keepdim=True)         # norm of the samples
    direction.div_(norm)                                      # unit-norm samples

    # sample norms from Gamma distribution:
    shape = torch.ones(mean.shape[1:]).mul_(mean.size(0))
    norm = torch.distributions.gamma.Gamma(shape, precision).sample()
    if norm.device != direction.device:
        norm = norm.to(direction.device)

    # return final sample:
    sample = direction.mul_(norm).add_(mean)
    if device is not None:
        sample = sample.to(device=device)
    return sample


install.py
#!/usr/bin/env python3

# Copyright (c) Facebook, Inc. and its affiliates.
# All rights reserved.

# This source code is licensed under the license found in the
# LICENSE file in the root directory of this source tree.

"""
Installs dependencies for open-source users.
"""

import os
import requests


# files for RDP dependency:
RDP_BASE = "https://raw.githubusercontent.com/tensorflow/privacy"
RDP_COMMIT = "1ce8cd4032b06e8afa475747a105cfcb01c52ebe"
RDP_FOLDER = "tensorflow_privacy/privacy/analysis"
RDP_FILES = {
    "rdp_accountant.py": "rdp_accountant.py",
    "compute_dp_sgd_privacy.py": "dpsgd_privacy.py"
}

# files for ResNet dependency:
RESNET_BASE = "https://raw.githubusercontent.com/akamaster/pytorch_resnet_cifar10"
RESNET_COMMIT = "4e4f8da1ba2611dad2eedf8a23505c0fbd94b983"
RESNET_FILES = {
    "resnet.py": "resnet.py",
}


def main():

    # download files needed for RDP accountant:
    for source, filename in RDP_FILES.items():
        url = f"{RDP_BASE}/{RDP_COMMIT}/{RDP_FOLDER}/{source}"
        request = requests.get(url, allow_redirects=True)
        with open(filename, "wb") as f:
            print(f"writing file {filename}")
            f.write(request.content)

    # download files needed for ResNet:
    for source, filename in RESNET_FILES.items():
        url = f"{RESNET_BASE}/{RESNET_COMMIT}/{source}"
        request = requests.get(url, allow_redirects=True)
        with open(filename, "wb") as f:
            print(f"writing file {filename}")
            f.write(request.content)

    # apply patches:
    for filename in RDP_FILES.values():
        patch_filename = filename.replace(".py", ".patch")
        os.system(f"patch {filename} patches/{patch_filename}")
    for filename in RESNET_FILES.values():
        patch_filename = filename.replace(".py", ".patch")
        os.system(f"patch {filename} patches/{patch_filename}")
    print("done.")


if __name__ == "__main__":
    main()


modules.py
#!/usr/bin/env python3

# Copyright (c) Facebook, Inc. and its affiliates.
# All rights reserved.

# This source code is licensed under the license found in the
# LICENSE file in the root directory of this source tree.

import math
import torch
import torch.nn as nn
import torch.nn.functional as F


class LinearFunction(torch.autograd.Function):
    """
    Linear function that computes gradients in a differentially private way.

    This uses https://arxiv.org/abs/1510.01799 to clip the per-example gradients
    without explicitly constructing them.
    """

    @staticmethod
    def forward(ctx, input, weight, bias, clip, std):

        # save context for backward:
        ctx.save_for_backward(input, weight, bias)
        ctx.clip = clip
        ctx.std = std

        # apply linear function and return:
        output = input.mm(weight.t())
        if bias is not None:
            output.add_(bias)
        return output

    @staticmethod
    def backward(ctx, grad_output):

        # unpack saved context:
        input, weight, bias = ctx.saved_tensors

        # compute gradient with respect to input (not private):
        grad_input = grad_output.mm(weight)

        # compute norms of per-example gradients with respect to parameters:
        gradient_norm = grad_output.pow(2.).sum(1, keepdim=True).mul(
            input.pow(2.).sum(1, keepdim=True)
        )
        gradient_norm = torch.sqrt(gradient_norm)

        # aggregate the clipped per-example gradients:
        multiplier = _get_multipliers(gradient_norm, ctx.clip)
        grad_weight = grad_output.div(multiplier + 1e-9).t().matmul(input)

        # add noise to gradient:
        grad_weight += torch.randn_like(grad_weight) * ctx.clip * ctx.std

        # perform same procedure for bias gradients:
        if bias is not None:
            multiplier = _get_multipliers(grad_output.norm(2, 1), ctx.clip)
            grad_bias = grad_output.mul(multiplier.unsqueeze(1)).sum(0)
            grad_bias += torch.randn_like(grad_bias) * ctx.clip * ctx.std
        else:
            grad_bias = None

        # return private gradients:
        return grad_input, grad_weight, grad_bias, None, None


class Conv2dFunction(torch.autograd.Function):
    """
    Conv2d function that computes gradients in a differentially private way.
    """

    @staticmethod
    def forward(ctx, input, weight, stride, padding, clip, std):

        # save context for backward:
        ctx.save_for_backward(input, weight)
        ctx.stride = stride
        ctx.padding = padding
        ctx.clip = clip
        ctx.std = std

        # perform convolution:
        return F.conv2d(
            input,
            weight,
            bias=None,
            stride=stride,
            padding=padding,
            dilation=1,
            groups=1,
        )

    @staticmethod
    def backward(ctx, grad_output):

        # get input, kernel, and sizes:
        input, weight = ctx.saved_tensors
        batch_size = input.size(0)
        out_channels, in_channels, weight_size_y, weight_size_x = weight.size()
        assert input.size(1) == in_channels, "wrong number of input channels"
        assert grad_output.size(1) == out_channels, "wrong number of output channels"
        assert grad_output.size(0) == batch_size, "wrong batch size"

        # compute gradient with respect to input:
        grad_input = torch.nn.grad.conv2d_input(
            input.size(),
            weight,
            grad_output,
            stride=ctx.stride,
            padding=ctx.padding,
        )

        # compute per-example gradient with respect to weights:
        out_channels, in_channels, weight_size_y, weight_size_x = weight.size()
        grad_output = grad_output.contiguous().repeat(1, in_channels, 1, 1)
        grad_output = grad_output.contiguous().view(
            grad_output.shape[0] * grad_output.shape[1],
            1,
            grad_output.shape[2],
            grad_output.shape[3],
        )
        input = input.contiguous().view(
            1,
            input.shape[0] * input.shape[1],
            input.shape[2],
            input.shape[3],
        )
        grad_weight = torch.conv2d(
            input,
            grad_output,
            None,
            1,
            ctx.padding,
            ctx.stride,
            in_channels * batch_size,
        )
        grad_weight = grad_weight.contiguous().view(
            batch_size,
            grad_weight.shape[1] // batch_size,
            grad_weight.shape[2],
            grad_weight.shape[3],
        )

        # compute norm of per-example weight gradients:
        grad_norm = torch.norm(
            grad_weight.view(batch_size, -1), p='fro', dim=1, keepdim=True,
        ).view(batch_size, 1, 1, 1)

        # aggregate the clipped per-example weight gradients:
        multiplier = _get_multipliers(grad_norm, ctx.clip)
        grad_weight = grad_weight.mul_(multiplier).sum(dim=0)
        grad_weight = grad_weight.view(
            in_channels,
            out_channels,
            grad_weight.shape[1],
            grad_weight.shape[2]
        ).transpose(0, 1).narrow(2, 0, weight_size_y).narrow(3, 0, weight_size_x)

        # add noise to gradient:
        grad_weight += torch.randn_like(grad_weight) * ctx.clip * ctx.std

        # return gradients:
        return grad_input, grad_weight, None, None, None, None


class AffineFunction(torch.autograd.Function):
    """
    Affine function that computes gradients in a differentially private way.
    """

    @staticmethod
    def forward(ctx, input, weight, bias, clip, std):

        # save context for backward pass:
        ctx.save_for_backward(input, weight, bias)
        ctx.clip = clip
        ctx.std = std

        # shape for broadcasting weights with input:
        broadcast_shape = [1] * input.dim()
        broadcast_shape[1] = input.shape[1]
        weight = weight.reshape(broadcast_shape)
        bias = bias.reshape(broadcast_shape)

        # apply affine transform to output before returning:
        output = input.mul(weight).add_(bias)
        return output

    @staticmethod
    def backward(ctx, grad_output):

        # unpack context from forward pass:
        input, weight, bias = ctx.saved_tensors

        # determine dimensions over which affine transform is broadcasted:
        stats_dimensions = list(range(grad_output.dim()))
        stats_dimensions.pop(1)

        # shape for broadcasting weights with grad_output:
        broadcast_shape = [1] * input.dim()
        broadcast_shape[1] = input.shape[1]
        weight = weight.reshape(broadcast_shape)
        bias = bias.reshape(broadcast_shape)

        # compute gradient with respect to input:
        grad_input = grad_output.mul(weight)

        # compute per-example gradient with respect to weights and biases:
        grad_weight = grad_output.mul(input)
        grad_bias = grad_output.clone()

        # compute norms of per-example gradients:
        batch_size = grad_output.size(0)
        grad_weight_norm = torch.norm(
            grad_weight.view(batch_size, -1), p='fro', dim=1, keepdim=True,
        )
        grad_bias_norm = torch.norm(
            grad_bias.view(batch_size, -1), p='fro', dim=1, keepdim=True,
        )

        # shape for broadcasting multipliers with grad_output:
        broadcast_shape = [1] * grad_output.dim()
        broadcast_shape[0] = grad_output.size(0)

        # aggregate the clipped per-example weight gradients:
        multiplier = _get_multipliers(grad_weight_norm, ctx.clip)
        multiplier = multiplier.reshape(broadcast_shape)
        grad_weight = grad_weight.mul_(multiplier).sum(stats_dimensions)

        # aggregate the clipped per-example weight gradients:
        multiplier = _get_multipliers(grad_bias_norm, ctx.clip)
        multiplier = multiplier.reshape(broadcast_shape)
        grad_bias = grad_bias.mul_(multiplier).sum(stats_dimensions)

        # add noise to gradients:
        grad_weight += torch.randn_like(grad_weight) * ctx.clip * ctx.std
        grad_bias += torch.randn_like(grad_bias) * ctx.clip * ctx.std

        # return gradients:
        return grad_input, grad_weight, grad_bias, None, None


# alias the application of the custom AutogradFunctions:
linear = LinearFunction.apply
conv2d = Conv2dFunction.apply
affine = AffineFunction.apply


class Linear(nn.Linear):
    """
    Extends nn.Linear to compute gradients in a differentially private way.
    Please refer to the documentation of nn.Linear for constructor signature.
    """

    def __init__(self, in_channels, out_channels, bias=True, clip=math.inf, std=0.0):
        super(Linear, self).__init__(in_channels, out_channels, bias=bias)
        self.clip = clip
        self.std = std

    def forward(self, input):
        if self.train:
            return linear(input, self.weight, self.bias, self.clip, self.std)
        else:
            return F.linear(input, self.weight, self.bias)


class Conv2d(nn.Conv2d):
    """
    Extends nn.Linear to compute gradients in a differentially private way.
    Please refer to the documentation of nn.Conv2d for constructor signature.
    This module does not support group convolution, dilation, or biases.
    """

    def __init__(
        self,
        in_channels,
        out_channels,
        kernel_size,
        stride=1,
        padding=0,
        bias=True,
        clip=math.inf,
        std=0.0,
    ):
        super(Conv2d, self).__init__(
            in_channels,
            out_channels,
            kernel_size,
            stride=stride,
            padding=padding,
            bias=bias,
        )
        self.clip = clip
        self.std = std
        assert self.bias is None, "Does not support bias in convolutions."
        assert self.groups == 1, "Does not support group convolutions."
        assert self.dilation == 1, "Does not support dilated convolutions."

    def forward(self, input):
        if self.train:
            return conv2d(
                input, self.weight, self.stride, self.padding, self.clip, self.std
            )
        else:
            return F.conv2d(
                input,
                self.weight,
                bias=self.bias,
                stride=self.stride,
                padding=self.padding,
                dilation=self.dilation,
                groups=self.groups,
            )


class GroupNorm(nn.GroupNorm):
    """
    Extends nn.GroupNorm to compute gradients in a differentially private way.
    Please refer to the documentation of nn.GroupNorm for constructor signature.
    """

    def __init__(
        self, num_groups, num_channels, eps=1e-5, affine=True, clip=math.inf, std=0.0
    ):
        super(GroupNorm, self).__init__(
            num_groups, num_channels, eps=eps, affine=affine
        )
        self.clip = clip
        self.std = std

    def forward(self, input):

        # training mode:
        if self.train:
            input_norm = F.group_norm(
                input, self.num_groups, None, None, self.eps
            )
            if self.weight is None or self.bias is None:
                return input_norm
            else:
                return affine(
                    input_norm, self.weight, self.bias, self.clip, self.std
                )  # only the affine transform needs private gradients

        # inference mode:
        else:
            return F.group_norm(
                input, self.num_groups, self.weight, self.bias, self.eps
            )


def _get_multipliers(gradient_norm, clip):
    """
    Given a vector of per-example `gradient_norm`s, computes the multipliers to
    be applied to the gradients to clip them at value `clip`.
    """
    multiplier = gradient_norm.new(gradient_norm.size()).fill_(1.)
    multiplier[gradient_norm.gt(clip)] = clip / gradient_norm[gradient_norm.gt(clip)]
    return multiplier


private_prediction_experiment.py
#!/usr/bin/env python3

# Copyright (c) Facebook, Inc. and its affiliates.
# All rights reserved.

# This source code is licensed under the license found in the
# LICENSE file in the root directory of this source tree.

import argparse
import copy
import json
import logging
import math
import os
import torch
import visdom

import dataloading
import private_prediction

# set up logger:
logger = logging.getLogger()
logger.setLevel(logging.INFO)

# hyperparameter values to try in cross-validation:
CROSS_VALIDATION = {
    "weight_decay": [1e-5, 5e-5, 1e-4, 5e-4, 1e-3, 5e-3, 1e-2, 5e-2, 1e-1, 5e-1],
    "clip": [5e-4, 1e-3, 2e-3, 5e-3, 1e-2, 2e-2, 5e-2, 1e-1, 2e-1, 5e-1],
}


def cross_validate(args, data, visualizer=None):
    """
    Performs cross-validation over hyperparameters for which this was requested.
    """

    # check if there are any parameters to cross-validate over:
    accuracies = {}
    arguments = {key: getattr(args, key) for key in CROSS_VALIDATION.keys()}
    if any(value == -1 for value in arguments.values()):

        # create validation split from training data:
        valid_size = data["train"]["features"].size(0) // 10
        original_train, data["valid"] = {}, {}
        for key in data["train"].keys():
            original_train[key] = data["train"][key]
            data["valid"][key] = original_train[key].narrow(0, 0, valid_size)
            data["train"][key] = original_train[key].narrow(
                0, valid_size, original_train[key].size(0) - valid_size
            )
        # NOTE: This assumes data is already shuffled.
        # NOTE: This makes an additional data copy, which may be bad on GPUs.

        # get hyperparameter key and values:
        hyper_key = [key for key, val in arguments.items() if val == -1]
        assert len(hyper_key) == 1, \
            "can only cross-validate over single hyperparameter at the same time"
        hyper_key = hyper_key[0]
        hyper_values = CROSS_VALIDATION[hyper_key]

        # perform the actual cross-validation:
        num_repetitions, idx = max(1, args.num_repetitions // 10), 0
        for hyper_value in hyper_values:

            # make copy of arguments that we can alter:
            args_copy = copy.deepcopy(args)
            setattr(args_copy, hyper_key, hyper_value)
            if args_copy.inference_budget == -1:
                args_copy.inference_budget = 100
            accuracies[hyper_value] = {}

            # repeat experiment multiple times:
            for _ in range(num_repetitions):
                logging.info(f"Cross-validation experiment {idx + 1} of "
                             f"{len(hyper_values) * num_repetitions}...")
                private_prediction.compute_accuracy(
                    args_copy, data,
                    accuracies=accuracies[hyper_value],
                    visualizer=visualizer,
                )
                idx += 1

        # find best hyperparameter setting:
        for hyper_value in hyper_values:
            valid_accuracy = accuracies[hyper_value]["valid"]
            if isinstance(valid_accuracy, dict):  # inference budget in accuracies
                valid_accuracy = valid_accuracy[str(args_copy.inference_budget)]
            accuracies[hyper_value] = sum(valid_accuracy) / float(num_repetitions)
        optimal_value = max(accuracies, key=accuracies.get)
        logging.info(f"Selecting {hyper_key} value of {optimal_value}...")

        # clean up validation set:
        data["train"] = original_train
        del data["valid"]

        # update arguments object:
        setattr(args, hyper_key, optimal_value)

    # return arguments to use for main experiment:
    return args


def main(args):
    """
    Runs private predictions experiment on dataset using input arguments `args`.
    """

    # set up visualizer:
    if args.visdom:
        visualizer = visdom.Visdom(args.visdom)
    if not args.visdom or not visualizer.check_connection():
        visualizer = None

    # load dataset:
    logging.info(f"Loading {args.dataset} dataset...")
    normalize = args.dataset.startswith("mnist")
    reshape = (args.model == "linear")
    num_classes = None if args.num_classes == -1 else args.num_classes
    data = {}
    for split in ["train", "test"]:
        data[split] = dataloading.load_dataset(
            name=args.dataset,
            split=split,
            normalize=normalize,
            reshape=reshape,
            num_classes=num_classes,
            root=args.data_folder,
        )

    # apply PCA if requested (on all data; non-transductive setting):
    if args.pca_dims != -1:
        assert reshape, "cannot use PCA with non-linear models"
        data["train"], mapping = dataloading.pca(data["train"], num_dims=args.pca_dims)
        data["test"], _ = dataloading.pca(data["test"], mapping=mapping)

    # subsample training data if requested:
    if args.num_samples != -1:
        data["train"] = dataloading.subsample(
            data["train"], num_samples=args.num_samples, random=False,
        )

    # copy data to GPU if requested (for linear models only):
    if args.device == "gpu" and args.model == "linear":
        assert torch.cuda.is_available(), "CUDA is not available on this machine."
        logging.info("Copying data to GPU...")
        for split in data.keys():
            for key, value in data[split].items():
                data[split][key] = value.cuda()

    # use cross-validation to tune hyperparameters:
    args = cross_validate(args, data, visualizer=visualizer)

    # repeat the same experiment multiple times:
    accuracies = {}
    for idx in range(args.num_repetitions):
        logging.info(f"Experiment {idx + 1} of {args.num_repetitions}...")
        private_prediction.compute_accuracy(
            args, data, accuracies=accuracies, visualizer=visualizer
        )

    # save results to file:
    if args.result_file is not None and args.result_file != "":
        logging.info(f"Writing results to file {args.result_file}...")
        with open(args.result_file, "wt") as json_file:
            json.dump(accuracies, json_file)


# run all the things:
if __name__ == '__main__':

    # parse input arguments:
    hostname = os.environ.get("HOSTNAME", "localhost")
    parser = argparse.ArgumentParser(description="Private prediction")
    parser.add_argument("--dataset", default="mnist", type=str,
                        help="dataset: 'mnist' (default), 'mnist1m', 'cifar10', "
                        "or 'cifar100'")
    parser.add_argument("--model", default="linear", type=str,
                        help="model: 'linear' (default) or "
                        "'resnet{20,32,44,56,110,1202}'")
    parser.add_argument("--method", default="subsagg", type=str,
                        help="private prediction method: 'subsagg' (default), "
                        "'loss_perturbation', '{model,logit}_sensitivity', or 'dpsgd'")
    parser.add_argument("--optimizer", default="lbfgs", type=str,
                        help="optimizer used for training: 'lbfgs' (default) or 'sgd'")
    parser.add_argument("--num_models", default=32, type=int,
                        help="number of models to train (for subsagg; default = 32)")
    parser.add_argument("--num_repetitions", default=10, type=int,
                        help="number of times to repeat experiment (default = 10)")
    parser.add_argument("--epsilon", default=math.inf, type=float,
                        help="privacy loss for predictions (default = infinity)")
    parser.add_argument("--delta", default=0.0, type=float,
                        help="privacy failure probability for predictions (default = 0.0)")
    parser.add_argument("--inference_budget", default=-1, type=int,
                        help="number of inferences to support (default = -1 for all)")
    parser.add_argument("--num_epochs", default=100, type=int,
                        help="number of training epochs (default = 100)")
    parser.add_argument("--batch_size", default=32, type=int,
                        help="batch size for SGD (default = 32)")
    parser.add_argument("--noise_dist", default="sqrt_gaussian", type=str,
                        help="noise distribution for sensitivity methods "
                        "methods: 'sqrt_gaussian' (default), 'laplacian', "
                        "'gaussian', 'advanced_gaussian'")
    parser.add_argument("--learning_rate", default=1., type=float,
                        help="initial learning rate for SGD (default = 1.0)")
    parser.add_argument("--weight_decay", default=0., type=float,
                        help="L2 regularization parameter (default = "
                        "0.0; set to -1 to cross-validate)")
    parser.add_argument("--clip", default=1e-1, type=float,
                        help="gradient clipping for DP-SGD method (default "
                        "= 1e-1; set to -1 to cross-validate)")
    parser.add_argument("--num_samples", default=-1, type=int,
                        help="number of training samples (default: all)")
    parser.add_argument("--num_classes", default=-1, type=int,
                        help="number of classes to use (default: all)")
    parser.add_argument("--pca_dims", default=-1, type=int,
                        help="number of PCA dimensions for data (default: not used)")
    parser.add_argument("--data_folder", default="/tmp", type=str,
                        help="folder in which to store data (default: '/tmp')")
    parser.add_argument("--result_file", default=None, type=str,
                        help="file in which to write experimental results")
    parser.add_argument("--device", default="cpu", type=str,
                        help="compute device: 'cpu' (default) or 'gpu'")
    parser.add_argument("--use_lr_scheduler", default=False, action="store_true",
                        help="use learning rate reduction (for DP-SGD)")
    parser.add_argument("--visdom", default=hostname, type=str,
                        help=f"visdom server (default = {hostname})")
    args = parser.parse_args()

    # check generic input arguments:
    if args.epsilon < 0.:
        raise ValueError("Epsilon value must be non-negative.")

    # run:
    main(args)
